app/uk/gov/hmrc/taxenrolmentassignmentfrontend/services/ThrottlingService.scala

```diff
@@ -0,0 +1,70 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.taxenrolmentassignmentfrontend.services
+
+import cats.data.EitherT
+import uk.gov.hmrc.auth.core.{Enrolment, EnrolmentIdentifier, Enrolments}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.service.TEAFResult
+import uk.gov.hmrc.taxenrolmentassignmentfrontend.AccountTypes
+import uk.gov.hmrc.taxenrolmentassignmentfrontend.AccountTypes.SINGLE_ACCOUNT
+import uk.gov.hmrc.taxenrolmentassignmentfrontend.config.AppConfig
+import uk.gov.hmrc.taxenrolmentassignmentfrontend.connectors.LegacyAuthConnector
+import uk.gov.hmrc.taxenrolmentassignmentfrontend.models.enums.EnrolmentEnum.hmrcPTKey
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+sealed trait ThrottleResult
+case object ThrottleApplied extends ThrottleResult
+case object ThrottleDoesNotApply extends ThrottleResult
+
+class ThrottlingService @Inject()(legacyAuthConnector: LegacyAuthConnector, appConfig: AppConfig) {
+
+  def throttle(accountType: AccountTypes.Value, nino: String, currentEnrolments: Set[Enrolment])(
+    implicit ec: ExecutionContext, hc: HeaderCarrier): TEAFResult[ThrottleResult] = {
+    if(shouldAccountTypeBeThrottled(accountType, appConfig.percentageOfUsersThrottledToGetFakeEnrolment, nino)) {
+      legacyAuthConnector.updateEnrolments(addPTEnrolmentToEnrolments(currentEnrolments, nino)).map(_ => ThrottleApplied)
+    } else {
+      EitherT.right(Future.successful(ThrottleDoesNotApply))
+    }
+  }
+
+  private[services] def addPTEnrolmentToEnrolments(currentEnrolments: Set[Enrolment], nino: String): Set[Enrolment] = {
+    currentEnrolments + newPTEnrolment(nino)
+  }
+
+  private[services] def newPTEnrolment(nino: String): Enrolment = {
+    Enrolment(s"$hmrcPTKey", Seq(EnrolmentIdentifier("NINO", nino)), "Activated", None)
+  }
+  private[services] def isNinoWithinThrottleThreshold(nino: String, percentageToThrottle: Int): Boolean = {
+    percentageToThrottle match {
+      case n if n >= 100 => false
+      case n if n < 0 => throw new IllegalArgumentException(s"percentage to throttle must be greater than 0, current percentage: $percentageToThrottle")
+      case _ if nino.length != 9 => throw new IllegalArgumentException(s"nino is incorrect length ${nino.length}")
+      case n => Try(nino.substring(6,8).toInt).map(ninoNumber => ninoNumber <= n).getOrElse(throw new IllegalArgumentException(s"nino was not valid format for throttle"))
+    }
+  }
+
+    private[services] def shouldAccountTypeBeThrottled(accountType: AccountTypes.Value,
+                                           percentageToThrottle: Int,
+                                           nino: String): Boolean = {
+    accountType != SINGLE_ACCOUNT && isNinoWithinThrottleThreshold(nino, percentageToThrottle)
```

Comment:
The account types that should be excluded from the throttle should also include PT_ASSIGNED_TO_CURRENT_USER and PT_ASSIGNED_TO_OTHER_USER

Reply:
Im thinking, maybe it should just also exclude PT_ASSIGNED_TO_CURRENT_USER
we want to include all other account types for now so it doesn't do exactly what we didn't want it to do, show the page for one NINO but not another.

Reply:
ill make this change, good spot

Reply:
done

Reply:
No it needs to include both PT_ASSIGNED_CURRENT_USER and PT_ASSIGNED_TO_OTHER_USER. The page around PT enrolment already being assigned we always want to show. I have had discussions with Kit Pimm around this and he agreed this should be the case. The only way the PT_ASSIGNED_TO_OTHER_USER can be displayed is if a user has been through the silent enrolment (either through us or the background update process) or completed the multiple accounts no enrolments or multiple accounts with SA on one of the accounts. For people with multiple accounts they will automatically be within the throttle, however the single account holders could potentially create a new account and then be allowed in if this page is throttled which we don't want. Once the enrolment has been assigned for real the user needs to sign in with that account or report the account.

Reply:
ok I see this was that edge case Kit mentioned

Reply:
Yes it is also the same reason the background process can't go first :-)